Integrated Jupiter Oracle for Token Pricing

luffy-orf opened the pull request:
## Integrated Jupiter Oracle for Token Pricing

### Issue resolved - #12 

#### This PR replaces the manual token price retrieval mechanism with Jupiter oracle integration to provide more accurate and real-time pricing. Key changes include:

- Implemented get_price function to fetch token prices from the Jupiter API.
- Added exponential backoff and retries for improved reliability.
- Enhanced error handling to handle API failures gracefully.
- Updated transaction logic to use the new price oracle.
- Added unit tests to validate the integration.

. 
 This pull request was created for https://app.gib.work/bounties/9429edb4-175f-40c5-bc29-eb57e79b3a71 in an attempt to solve a bounty #12 . Payment for the bounty is immediately sent to the contributor after merge.

pkxro commented:
This generally looks good - please run the linter for CI

luffy-orf commented:
@pkxro Could you review the code and let me know if anything needs to be changed

gitteri commented on crates/lib/src/transaction/fees.rs:
```diff
@@ -78,15 +79,18 @@ pub async fn calculate_token_value_in_lamports(
+
     let sol_per_token =
-        price_info.price * LAMPORTS_PER_SOL as f64 / (10f64.powi(mint_data.decimals as i32));
+        token_price.price * LAMPORTS_PER_SOL as f64 / (10f64.powi(mint_data.decimals as i32));
```
The logic in this function is off. You need to get both the mint and sol prices in order to convert between the two (we should not assume the mint is a stablecoin). Note that in the [Jup API](https://station.jup.ag/docs/old/apis/price-api-v2#usage) you can get more than one token price at a time.

Here's the general functionality to go from a payment token amount to lamports:
```
pub fn convert_token_to_lamports(
    token_amount: u64,
    token_decimals: u8,
    token_price_usd: f64,
    sol_price_usd: f64,
) -> u64 {
    // SOL decimals (1 SOL = 10^9 lamports)
    const SOL_DECIMALS: u8 = 9;
    
    // Step 1: Convert raw token amount to actual token amount
    let actual_token_amount = token_amount as f64 / (10u64.pow(token_decimals as u32) as f64);
    
    // Step 2: Calculate the USD value of the token amount
    let usd_value = actual_token_amount * token_price_usd;
    
    // Step 3: Convert USD value to SOL
    let sol_amount = usd_value / sol_price_usd;
    
    // Step 4: Convert SOL to lamports
    let lamports = sol_amount * (10u64.pow(SOL_DECIMALS as u32) as f64);
    
    // Round down to get a whole number of lamports
    lamports.floor() as u64
}
```

luffy-orf replied:
Updated the function to correctly factor in both the token and SOL prices for accurate conversion. Used the suggested approach and ensured proper formatting. Let me know if any further adjustments are needed!

gitteri commented on crates/rpc/src/method/sign_transaction.rs:
```diff
@@ -28,7 +28,7 @@ pub async fn sign_transaction(
 ) -> Result<SignTransactionResponse, KoraError> {
     let encoding = request.encoding.unwrap_or_default();
     let transaction = encoding.decode_transaction(&request.transaction)?;
-    let (transaction, signed_transaction) =
+    let (transaction, _signed_transaction) =
```
undo this change please

luffy-orf replied:
Done

gitteri commented on crates/lib/src/transaction/fees.rs:
```diff
@@ -78,17 +79,37 @@ pub async fn calculate_token_value_in_lamports(
+    let sol_price = oracle.get_token_price("SOL").await?;
+
+    // SOL has a fixed decimal representation (1 SOL = 10^9 lamports)
+    const SOL_DECIMALS: u8 = 9;
```
You can use the constant from solana sdk instead: https://docs.rs/solana-native-token/2.2.1/solana_native_token/constant.LAMPORTS_PER_SOL.html

luffy-orf replied:
Updated to use LAMPORTS_PER_SOL from Solana SDK

gitteri commented on crates/lib/src/transaction/fees.rs:
```diff
@@ -78,17 +79,37 @@ pub async fn calculate_token_value_in_lamports(
+    let oracle = PriceOracle::new(3, Duration::from_secs(1));
 
-    Ok(lamport_value)
+    // Fetch token price in USD
+    let token_price = oracle.get_token_price(&mint.to_string()).await?;
+
+    // Fetch SOL price in USD (required for conversion)
+    let sol_price = oracle.get_token_price("SOL").await?;
```
be sure to return the error if there is one. My comment was just example code

luffy-orf replied:
Added error handling

gitteri commented on crates/lib/src/oracle/jupiter.rs:
```diff
@@ -0,0 +1,43 @@
+use reqwest::Client;
+use serde::Deserialize;
+
+const JUPITER_API_URL: &str = "https://price.jup.ag/v4";
```
Jupiter has a new host name for their api: https://api.jup.ag/price/v2
https://station.jup.ag/docs/

luffy-orf replied:
updated the url

luffy-orf commented on crates/rpc/src/method/sign_transaction_if_paid.rs:
```diff
@@ -1,51 +1,41 @@
-    rpc_client: &Arc<RpcClient>,
-    validation: &ValidationConfig,
     request: SignTransactionIfPaidRequest,
+    rpc_client: &RpcClient,
+    validation: &ValidationConfig,
```
I initially reordered the parameters while making other changes, but I realized that it’s causing type mismatches elsewhere in the code. Since the original ordering was correct and expected in function calls, I'll revert it to the previous order to avoid breaking anything. Thanks for pointing it out!

luffy-orf commented:
@gitteri I've addressed all the requested changes. Let me know if there's anything else I should improve. Appreciate your feedback!

luffy-orf commented:
@gitteri all requested changes have been made, and tests have passed. Looking forward to your approval. Thanks for your feedback and review!

luffy-orf commented on crates/cli/src/main.rs:
```diff
@@ -132,16 +132,10 @@ async fn main() -> Result<(), KoraError> {
-                token_price_info,
-            )
-            .await?;
+            let _token_price_info = token_price.map(|price| TokenPriceInfo { price });
```
Removed that line

luffy-orf commented:
@gitteri  All requested changes have been implemented. I’ve removed the unnecessary line, and simplified the price fetching logic using Jupiter’s vsToken query parameter as suggested. Let me know if anything else needs adjustment!

luffy-orf commented on crates/lib/src/transaction/fees.rs:
```diff
@@ -78,17 +79,43 @@ pub async fn calculate_token_value_in_lamports(
+    let usd_value = token_amount * token_price.price;
+
+    // Convert USD value to equivalent SOL amount
+    let sol_amount = usd_value / sol_price.price;
```
Done

gitteri commented on crates/lib/src/oracle/tests.rs:
```diff
@@ -0,0 +1,46 @@
+use tokio;
+
+#[tokio::test]
+async fn test_jupiter_price_fetch() {
```
This is an example response. please update these test to match this this response format
```
{"data":{"So11111111111111111111111111111111111111112":{"id":"So11111111111111111111111111111111111111112","type":"derivedPrice","price":"1"},"JUPyiwrYJFskUPiHa7hkeR8VUtAeFoSYbKedZNsDvCN":{"id":"JUPyiwrYJFskUPiHa7hkeR8VUtAeFoSYbKedZNsDvCN","type":"derivedPrice","price":"0.005321503266927636"}},"timeTaken":0.003297425}
```

luffy-orf replied:
@gitteri  I've updated the test mock responses to match the exact Jupiter API response format you provided. The tests now properly validate against the correct data structure including the type field, string prices, and timeTaken field.

luffy-orf commented:
@gitteri Tests have passed. Looking forward to your approval. Thanks for your feedback and review!
